Apps/Laboratory/src/Services/LabTestKitService.cs

```diff
@@ -0,0 +1,153 @@
+        /// <param name="labTestKitClient">The client to use for lab tests.</param>
+        public LabTestKitService(
+            ILogger<LabTestKitService> logger,
+            IHttpContextAccessor httpContextAccessor,
+            ITokenCacheService tokenCacheService,
+            ILabTestKitClient labTestKitClient)
+        {
+            this.logger = logger;
+            this.httpContextAccessor = httpContextAccessor;
+            this.tokenCacheService = tokenCacheService;
+            this.labTestKitClient = labTestKitClient;
+        }
+
+        /// <inheritdoc/>
+        public async Task<RequestResult<PublicLabTestKit>> RegisterLabTestKitAsync(PublicLabTestKit testKit)
+        {
+            RequestResult<PublicLabTestKit> requestResult = InitializeResult<PublicLabTestKit>(testKit);
+
+            // Use a system token
+            string? accessToken = this.tokenCacheService.RetrieveAccessToken();
+            if (accessToken != null)
+            {
+                HttpResponseMessage response =
+                    await this.labTestKitClient.RegisterLabTest(testKit, accessToken).ConfigureAwait(true);
+                ProcessResponse<PublicLabTestKit>(requestResult, response);
+            }
+            else
+            {
+                requestResult.ResultError = new RequestResultError()
+                {
+                    ResultMessage = $"Unable to acquire authentication token",
+                    ErrorCode = ErrorTranslator.ServiceError(ErrorType.CommunicationExternal, ServiceType.Keycloak),
+                };
+                this.logger.LogError("Unable to acquire authentication token");
+            }
+
+            return requestResult;
+        }
+
+        /// <inheritdoc/>
+        public async Task<RequestResult<LabTestKit>> RegisterLabTestKitAsync(string hdid, LabTestKit testKit)
+        {
+            RequestResult<LabTestKit> requestResult = InitializeResult<LabTestKit>(testKit);
+
+            // Use the users token
+            HttpContext? httpContext = this.httpContextAccessor.HttpContext;
+            string? accessToken = await httpContext.GetTokenAsync("access_token").ConfigureAwait(true);
+            HttpResponseMessage response =
+                await this.labTestKitClient.RegisterLabTest(hdid, testKit, accessToken).ConfigureAwait(true);
+            ProcessResponse<LabTestKit>(requestResult, response);
+
+            return requestResult;
+        }
+
+        private static RequestResult<T> InitializeResult<T>(T payload)
+            where T : class
+        {
+            RequestResult<T> result = new()
+            {
+                ResourcePayload = payload,
+                TotalResultCount = 0,
+                ResultStatus = ResultType.Error,
+            };
+
+            return result;
+        }
+
+        private static void ProcessResponse<T>(RequestResult<T> requestResult, HttpResponseMessage response)
```

Comment:
Same as line 100.  Is it possible to move this private static method into a common class with a more generic method so that all services can use it.  This would definitely reduce the amount of if logic in the service.

Reply:
same answer again, this is custom to this service and is really the business logic.

Reply:
As discussed this code has now been integrated into the common auth delegate.